### PR TITLE
Cache stats by saving url to db

### PR DESCRIPTION
**Pull request by xjahic**

### CODE REVIEW

File: src/cogs/admin_commands.py [hunk 1]
```diff
@@ -175,3 +175,14 @@ async def send_survey(self, ctx, survey_id):
                 f"Calling send-survey command for survey_id {survey_id}, but this survey doesn't exist"
             )
             await ctx.channel.send(f"Survey with id {survey_id} wasn't found")
+
+    @commands.command(name="clear-show-stats-cache")
+    @commands.has_role(ADMIN_ROLE_ID)
+    async def clear_show_stats_cache(self, ctx):
+        """
+        Only member with admin role can run this command
```
Comment (gabrielKerekes) [inline src/cogs/admin_commands.py:183]: I'd add periods to the text as they are pretty much sentences.

File: src/stats/show_stats.py [hunk 2]
```diff
@@ -37,6 +42,16 @@ async def show_stats(ctx, stat_id):
...
     survey_question_id, stat_type, stat_title = stats_options[stat_id]
 
+    if cached_stat := db.get_show_stats_updated_at_with_url_or_none(
+        stat_id
+    ):
+        updated_at, url = cached_stat
+        if not is_graph_expired(updated_at):
+            await send_graph_to_channel(ctx, url, stat_title)
+            return
+        else:
```
Comment (gabrielKerekes) [inline src/stats/show_stats.py:52]: `else` is IMHO redundant.

File: src/stats/show_stats.py [hunk 3]
```diff
@@ -49,8 +64,10 @@ async def show_stats(ctx, stat_id):
         make_percentage_graph(
             stat_id, options, answers, stat_title, graph_size=graph_size
         )
-
-        await send_graph_to_channel(ctx, stat_id, stat_title)
+        if imgur_url := await upload_graph_to_imgur_or_none(ctx, stat_id):
+            db.add_show_stats_cache(stat_id, imgur_url)
+            await send_graph_to_channel(ctx, imgur_url, stat_title)
+        return
```
Comment (gabrielKerekes) [inline src/stats/show_stats.py:70]: `return` is not needed.

File: src/stats/show_stats.py [hunk 4]
```diff
@@ -49,8 +64,10 @@ async def show_stats(ctx, stat_id):
         make_percentage_graph(
             stat_id, options, answers, stat_title, graph_size=graph_size
         )
-
-        await send_graph_to_channel(ctx, stat_id, stat_title)
+        if imgur_url := await upload_graph_to_imgur_or_none(ctx, stat_id):
```
Comment (gabrielKerekes) [inline src/stats/show_stats.py:67]: This whole `if` could be moved out of the `PERCENTAGE_BAR` if. But maybe we can worry about that later.

File: src/stats/show_stats.py [hunk 5]
```diff
@@ -113,12 +131,24 @@ def print_percentages_above_graph_bars(graph, data):
...
+    graph_embed.set_image(url=url)
     await ctx.channel.send(embed=graph_embed)
 
 
-def upload_graph_to_imgur(stat_id):
-    uploaded_image = imgur.upload_image(f"{stat_id}.png")
-    return uploaded_image.link
+async def upload_graph_to_imgur_or_none(ctx, stat_id):
+    try:
+        uploaded_image = imgur.upload_image(f"{stat_id}.png")
```
Comment (gabrielKerekes) [inline src/stats/show_stats.py:142]: How come we can store the `{stat_id}.png` file now? Or is it just in memory? 🤷

File: src/db/__init__.py [hunk 6]
```diff
@@ -21,6 +21,20 @@ def get_survey_question_order_or_none(question_id):
     return _get_first_or_none(question)
 
 
+def get_show_stats_updated_at_with_url_or_none(stat_id):
+    row = _fetchone(
+        f"SELECT updated_at, url FROM show_stats_cache WHERE stat_id = {stat_id}"
+    )
+    if row is None:
+        return None
+    else:
```
Comment (gabrielKerekes) [inline src/db/__init__.py:30]: `else` not needed.

File: src/stats/show_stats.py [hunk 7]
```diff
@@ -113,12 +131,24 @@ def print_percentages_above_graph_bars(graph, data):
...
+        return uploaded_image.link
+    # Imgur occasionally responds with 403 because their servers are overloaded
+    # there is nothing to do (except buy premium)
+    except HTTPError:
+        await ctx.channel.send("Server preťažený, skús prosím neskôr.")
```
Comment (gabrielKerekes) [inline src/stats/show_stats.py:147]: This seems to me like an unexpected side-effect of the function. Maybe we can just return the `None` and send the message in callee. Yes we kind of lose the information about the type of the error, but we can simply always send the same messages no matter what the error is.

**Review by gabrielKerekes — CHANGES_REQUESTED** [review]
Please address the comments. And fix style checks 😉

File: src/stats/show_stats.py (hunk 2 above)
Comment (xjahic) [reply src/stats/show_stats.py:52]: I will leave it there for better readibility

File: src/stats/show_stats.py (hunk 5 above)
Comment (xjahic) [reply src/stats/show_stats.py:142]: Storing files works only for short periods of time. 
https://help.heroku.com/K1PPS2WM/why-are-my-file-uploads-missing-deleted

**Pull request merged**